caliban/hooks/util.py

```diff
@@ -0,0 +1,83 @@
+#!/usr/bin/python
+#
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Functions to execute hooks
+"""
+
+from __future__ import absolute_import, division, print_function
+
+from typing import Any, Dict, List, NamedTuple, NewType, Optional, Union
+
+import subprocess
+from absl import logging
+import json
+
+import caliban.util.fs as ufs
+
+def perform_prebuild_hooks(caliban_config: Dict[str, Any]) -> Dict:
```

Comment:
I'd recommend a function that calls a single hook and returns its error OR successful dict. (Also, echoing Ambrose, let's get some more descriptive types on here!)

And then the final function can be a multi-version of that function, and call ALL of the hooks, bailing out at the first failure or collecting all of the results.

Oh man, we've re-discovered `Traverse` from functional programming: https://typelevel.org/cats/typeclasses/traverse.html, https://eed3si9n.com/herding-cats/Traverse.html

and the applicative functor on `Either`!!

Crazily this actually is a really useful idea, the `Validated` type. https://eed3si9n.com/herding-cats/Validated.html

This lets you take a bunch of possibly-failing things like this, or like, say, validations of stuff in a web form, and:

- run ALL of the validations, and
- return ALL of the errors

the easiest thing to write is what we want here, where you bail at the first error. But it can be nice to imagine that you want to execute hooks in parallel and return all errors.